### ISSUE TITLE
Enhancement: [no-unnecessary-condition] report on type guards usages where the type is already equal

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10007]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

Report when checking type guards, where the type of the variable is already _exactly_ equal to the type guard return type.

### Fail

```TypeScript
function isBool(x: unknown): x is boolean {
    return typeof x === 'boolean';
}

declare const b: boolean;

if (isBool(b)) { // this is unnecessary, since b is already boolean.
}
```

### Pass

```TypeScript
declare const maybeBoolean: boolean | string;

if (isBool(maybeBoolean)) { // necessary - might not a boolean.
}

declare const t: true;

if (isBool(t)) { // debatable - t is a subtype of boolean, not exactly equal. this has some funny effects in the type system.
    console.log(t); // t gets widened to boolean here, even though it was typed as true.
} else {
    console.error(t); // t also widened to boolean here.
}
```

### Additional Info

https://github.com/typescript-eslint/typescript-eslint/issues/9076 is in a similar domain, checking unnecessary assertion functions.

**Comment by kirkwaiblinger** [comment]
Thinking about this, I think that assertion functions should be treated the same:
```ts
function assertBool(x: unknown): asserts x is boolean {
    if (typeof x !== 'boolean') throw new Error();
}

declare const b: boolean;

assertBool(b); // unnecessary; already boolean
```

(very analogous to #9076)

**Comment by kirkwaiblinger** [comment]
Wondering if these should also be under their own options, or whether they should share the option being created in https://github.com/typescript-eslint/typescript-eslint/pull/10009
